_posts/2025-06-24-kafka-queue.md

```diff
@@ -0,0 +1,590 @@
+
+Traditional message queue and Kafka topic are significantly different in terms of design and use cases. Queue allows a group of consumers to read and process records in parallel to distribute the workload. It does not typically provide a strong ordering guarantee as records in a queue are consumed and processed independently. However it does provide exactly-once guarantee as each record is only read once and is deleted from the queue. States of individual records are usually tracked and acknowledged after being processed successfully to avoid prematurely removing records off the queue or to handle unprocessable records. This also means that records in a queue are not replayable, therefore a queue can only be consumed by a single group of consumers.
+
+A Kafka topic on the other hand, can be consumed by multiple groups of consumers as it is an immutable log and reading of records does not result in removal. It also provides an ordering mechanism by using record keys. However, workload distribution cannot be achieved alone with consumers. A topic has to be partitioned in order to spread the workload across a group of consumers.
+
+The Queues for Kafka feature allows you to use a Kafka topic like a traditional message queue and increase parallelism of message processing beyond the number of partitions. It provides queue-like semantics through a new consumer group called <b>share group</b> to enable use cases for traditional message queue. This new type of group gives more fine-grained control on message acknowledgement and retries.
+
+The key difference between share group and regular consumer group is how partitions get assigned to consumer members. With regular consumer groups, each partition is exclusively assigned to a single member of the consumer group. Therefore users typically can have as many consumer members as the number of partitions to maximise the parallelism in message processing. Moreover, due to this, users tend to over partition their topics in order to cope with peak loads that may only happen sometimes. However, share groups balance partitions between all members of a share group, allowing multiple consumer members to fetch from the same partition. So users can have more consumers than the number of partitions, further increasing the parallelism and they do not need to over partition their topics, but can just scale up and down their consumers to cope with the peak loads. When share group members consume from the same partition, each record on that partition is still only read by one consumer in the group.
+
+The share group currently does not provide ordering guarantee as multiple consumers can fetch records from the same partition. And it does not support dead letter queue to handle unprocessable/poison records. However, these features may potentially be added in the future.
```

Comment:
Should we also mention the share group doesn't support exactly-once semantics now given we've mentioned it above? In the [KIP](https://cwiki.apache.org/confluence/display/KAFKA/KIP-932%3A+Queues+for+Kafka#KIP932:QueuesforKafka-FutureWork), it said:

> Finally, this KIP does not include support for acknowledging delivery using transactions for exactly-once semantics.